plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -41,20 +46,71 @@ abstractly specifiable. It's an implementation detail.
 
 -}
 
+{- Note [Integer types for costing]
+We care about the speed of our integer operations for costing, this has a significant effect on speed.
+But we also need to care about overflow: the cost counters overflowing is a potential attack!
+
+We have a few choices here for what to do with an overflow:
+- Don't (this is what 'Integer' does, it's unbounded)
+- Wrap (this is what 'Int'/'Int64' and friends do)
+- Throw an overflow error (this is what 'Data.SafeInt' does)
+- Saturate (i.e. return max/min bound, this is what 'Data.SatInt does)
+
+In our case
+- Not overflowing would be nice, but 'Integer' is significantly slower than the other types.
+- Wrapping is quite dangerous, as it could lead to us getting attacked by someone wrapping
+their cost around to something that looks below the budget.
+- Throwing would be okay, but we'd have to worry about exception catching.
+- Saturating is actually quite nice: we care about whether `a op b < budget`. So long as `budget < maxBound`,
+  then `a op b < budget` will have the same truth value *regardless* of whether the operation overflows and saturates,
+  since saturating implies `a op b >= maxBound > budget`. Plus, it means we don't need to deal with
+  exceptions.
+
+So we use 'Data.SatInt', a variant of 'Data.SafeInt' that does saturating arithmetic.
+
+'SatInt' is quite fast, but not quite as fast as using 'Int64' directly (I don't know why that would be, apart from maybe
+just the overflow checks), but the wrapping behaviour of 'Int64' is unacceptable..
+
+One other wrinkle is that 'SatInt' is backed by an 'Int' (i.e. a machine integer with platform-dependent
+size), rather than an 'Int64' since the primops that we need are only available for 'Int' until GHC 9.2
+or so. So on 32bit platforms, we have much less header
+
+However we mostly care about 64bit platforms, so this isn't too much of a problem. The only one where it could be a problem
+is GHCJS, which does present as a 32bit platform. However, we won't care about *performance* on GHCJS, since
+nobody will be running a node compiled to JS (and if they do, they deserve terrible performance). So:
+if we are not on a 64bit platform, then we can just fallback to the slower (but safe) 'Integer'.
+
+-}
+
+-- See Note [Integer types for costing]
+type CostingInteger =
+#if WORD_SIZE_IN_BITS < 64
+    Integer
+#else
+    SatInt
+#endif
+
+-- $(if finiteBitSize (0::SatInt) < 64 then [t|Integer|] else [t|SatInt|])
+
 -- | Counts size in machine words (64bit for the near future)
-newtype ExMemory = ExMemory Integer
+newtype ExMemory = ExMemory CostingInteger
   deriving (Eq, Ord, Show)
-  deriving newtype (Num, Pretty, NFData)
-  deriving (Semigroup, Monoid) via (Sum Integer)
-deriving newtype instance PrettyDefaultBy config Integer => PrettyBy config ExMemory
+  deriving newtype (Num, NFData)
+  deriving (Semigroup, Monoid) via (Sum CostingInteger)
+instance Pretty ExMemory where
+    pretty (ExMemory i) = pretty (toInteger i)
```

Comment:
If I invert the condition in the `#if` then I get a number of (entirely reasonable) warnings like this.

```
    Call of toInteger :: CostingInteger -> Integer
      can probably be omitted
    |
109 |     pretty (ExMemory i) = pretty (toInteger i)
    |                                   ^^^^^^^^^
```
I don't know what can be done about this other than enclosing these instances in a `#if` as well, which is surely not good.  Will the false case of the `#if` ever be tested in CI (I'm guessing not)?

Reply:
Yeah, I tried that and I figured the warnings were okay. We could suppress them "just in case", but it's not a problem unless you have `-Werror`.